src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```diff
@@ -0,0 +1,748 @@
+               get_name(),
+               state);
+    const auto& [status, status_msg] = HCS().pause_compute_system(hcs_system);
+    set_state(fetch_state_from_api());
+    handle_state_update();
+}
+
+HCSVirtualMachine::State HCSVirtualMachine::current_state()
+{
+    return state;
+}
+int HCSVirtualMachine::ssh_port()
+{
+    return default_ssh_port;
+}
+std::string HCSVirtualMachine::ssh_hostname(std::chrono::milliseconds /*timeout*/)
+{
+    return fmt::format("{}.mshome.net", get_name());
+}
+std::string HCSVirtualMachine::ssh_username()
+{
+    return description.ssh_username;
+}
+
+std::optional<IPAddress> HCSVirtualMachine::management_ipv4()
+{
+    const auto& [ipv4, _] = resolve_ip_addresses(ssh_hostname({}).c_str());
+    if (ipv4.empty())
+    {
+        mpl::error(log_category, "management_ipv4() > failed to resolve `{}`", ssh_hostname({}));
+        return std::nullopt;
+    }
+
+    const auto result = *ipv4.begin();
+
+    mpl::trace(log_category, "management_ipv4() > IP address is `{}`", result);
+
+    // Prefer the first one
+    return std::make_optional<IPAddress>(result);
+}
+
+void HCSVirtualMachine::handle_state_update()
+{
+    monitor.persist_state_for(get_name(), state);
+}
+
+hcs::ComputeSystemState HCSVirtualMachine::fetch_state_from_api()
+{
+    hcs::ComputeSystemState compute_system_state{hcs::ComputeSystemState::unknown};
+    const auto result = HCS().get_compute_system_state(hcs_system, compute_system_state);
+    return compute_system_state;
+}
+
+void HCSVirtualMachine::update_cpus(int num_cores)
+{
+    mpl::debug(log_category,
+               "update_cpus() -> called for VM `{}`, num_cores `{}`",
+               get_name(),
+               num_cores);
+    description.num_cores = num_cores;
+}
+
+void HCSVirtualMachine::resize_memory(const MemorySize& new_size)
+{
+    mpl::debug(log_category,
+               "resize_memory() -> called for VM `{}`, new_size `{}` MiB",
+               get_name(),
+               new_size.in_megabytes());
+    description.mem_size = new_size;
+
+    // The implementation below allows runtime resize of memory. Multipass currently does
+    // not support runtime resize so commenting it out.
+
+    // hcs::HcsRequest req{
+    //     hcs::HcsResourcePath::Memory(),
+    //     hcs::HcsRequestType::Update(),
+    //     hcs::HcsModifyMemorySettings{static_cast<std::uint32_t>(new_size.in_megabytes())}};
+    // HCS().modify_compute_system(hcs_system, req);
+}
+
+void HCSVirtualMachine::resize_disk(const MemorySize& new_size)
+{
+    mpl::debug(log_category,
+               "resize_disk() -> called for VM `{}`, new_size `{}` MiB",
+               get_name(),
+               new_size.in_megabytes());
+
+    if (get_num_snapshots() > 0)
+    {
+        throw ResizeDiskWithSnapshotsException{"Cannot resize the primary disk while there are "
+                                               "snapshots. To resize, delete the snapshots first."};
```

Comment:
Interesting. I know this is the official recommendation. However, growing snapshot layers (with Ubuntu) always worked in all our testing, so we chose to allow it in the old Hyper-V. And we never had a user complaint. 

Did you actually try resizing the head layer?

Reply:
No, I haven't tried that but I don't think we should play russian roulette when it comes to user data. There are block sizes, offsets and metadata to be re-calculated and our snapshot implementation does not do any of that at the moment (maybe hyper-v is doing so in the background). Is it a crucial use-case to support? If so, I can look into supporting this properly.